#' NOMeseq data for WT and AdnpKO mouse ES cells
#'
#' WT and AdnpKO mouse ES cells were subjected to guided NOMe-seq.
#' 1500 regions were targeted using guideRNAs and Cas9, and bisulfite
#' sequenced in 300bp paired-end mode. Reads were mapped to the genome
#' using biscuit and the GCH protection was determined using the nomeR package.
#' To reduce file size, rows of the nomeR output were filtered to contain 30-500 fragments.
#'
#' @format ## `NomeMatrix`
#' A tibble with 1397 rows and 4 columns:
#' \describe{
#'   \item{SampleName}{sample names}
#'   \item{names}{ROI names with the format: TFmotif_chromosome_start_end}
#'   \item{nFragsAnalyzed}{number of fragments that were analyzed for GCH methylation protection}
#'   \item{GCH_DataMatrix}{matrix with a column for every position around a motif center and a row for every fragment,
#'   numbers indicating GCH methylation protection where 1 = protected and 0 = not protected}
#' }
#' @source generated by Lucas Kaaij
#'
#' @return A tibble with 1397 rows and 4 columns
"NomeMatrix"
