---
title: "dinoR-vignette"
author: " Michaela Schwaiger"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{dinoR-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

#### load the necessary packages

```{r setup}
suppressPackageStartupMessages({
  library(dinoR)
  library(ggplot2)
  library(dplyr)
  library(SummarizedExperiment)
  })
```

### load the NomeMatrix data for 2 replicates Adnp_KO and WT

We use biscuit to map 300bp paired-end reads to the genome, umi_tools to remove duplicated UMIs, and the get_data_matrix_from_bams function from the nomeR package (E. Ozonov, unpublished) to get the protection from GCH methylation calls for each read pair (fragment) overlapping a region of interest (ROI). However, any tool that will result in GCH methylation protection calls can be used (for example QuasR), as long as the final data structure is as shown below. Note that we use protection from methylation calls (0 = methylated, 1 = not methylated). 

```{r load data}
NomeData <- readRDS(system.file("extdata", "NOMeSeqData.rds", package = "dinoR"))
NomeData
```

### meta plots across ROIs with common TF motifs in the center

We generate metaplots, grouping our ROIs into those that have  Rest, Ctcf, or Adnp motif sites in their center.
We use 2 samples from WT mouse ES cells,a nd two samples from Adnp KO mouse ES cells.

```{r,fig.height=6,fig.width=8}
avePlotData <- metaPlots(NomeData=NomeData,nr=10,ROIgroup = "motif")

#plot average plots
ggplot(avePlotData, aes(x=position,y=protection)) + geom_point(alpha=0.5) + 
    geom_line(aes(x=position,y=loess),col="darkblue",lwd=2) + 
    theme_classic() + facet_grid(rows = vars(type),cols= vars(sample), scales = "free") +
    ylim(c(0,100)) + geom_hline(yintercept = c(10,20,30,40,50,60,70,80,90),
                                alpha=0.5,color="grey",linetype="dashed")
```

We can already see that while the NOMe footprints around Rest and Ctcf motifs don't change, there are clear differences around the Adnp motifs.

To quantify these differences, we can now classify each fragment according to five types of footprints: transcription factor bound (TF), open chromatin, and nucleosome (we distinguish also upstream nucleosome (upNuc), downstream nucleosome (downNuc), and all other nucleosome (Nuc) footprints). To do this we use three windows (-50:-25, -8:8, 25:50) around the ROI center. Then we count the number of fragments in each sample-ROI combination supporting each footprint category. 

### generate read counts for TF open nuc 

```{r read counts}
NomeData <- footprintCalc(NomeData)
NomeData <- footprintQuant(NomeData)
#remove any rows where no reads could be assigned a footprint
#footprint_counts <- footprint_counts[rowSums(footprint_counts[,3:7]) > 0,]
NomeData
```

Note that if a fragment does not have methylation protection data in all three windows needed for classification, the fragment will be counted as noData. 

Next we can test for differential abundance of footprints between Adnp KO and WT samples.

### calculate differential NOMe-seq footprint abundance between Adnp KO and WT 

```{r diNOMe}
res <- diNOMeTest(NomeData,WTsamples = c("WT_1","WT_2"),
  KOsamples = c("AdnpKO_1","AdnpKO_2"))
res
```

#### plot the number of regulated ROIs within each ROI type

```{r nregulated, fig.height=4,fig.width=7}

res %>% group_by(contrasts,motif,regulated) %>% summarize(n=n()) %>% 
  ggplot(aes(x=motif,y=n,fill=regulated)) + geom_bar(stat="identity") + 
  facet_grid(~contrasts) + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  scale_fill_manual(values=c("orange","grey","blue3"))
```


#### MA plots

```{r MAplot,fig.height=2,fig.width=10}
ggplot(res,aes(y=logFC,x=logCPM,col=regulated)) + geom_point() + 
  facet_grid(~contrasts) + theme_bw() + 
  scale_color_manual(values=c("orange","grey","blue3"))
```

### calculate the percentage of fragments in each footprint type and plot a heatmap comparing percentages in WT and KO

```{r percentages,fig.height=5,fig.width=7}
footprint_percentages <- footprintPerc(NomeData)

#plot heatmaps 
fpPercHeatmap(footprint_percentages)
```

### compare the footprint percentages and significance testing results for KO and WT 

```{r comparison,fig.height=4,fig.width=8}

compareFootprints(footprint_percentages,res,WTsamples = c("WT_1","WT_2"),
                  KOsamples = c("AdnpKO_1","AdnpKO_2"),plotcols = c("#f03b20", "#a8ddb5", "#bdbdbd"))

```

We can see that in Adnp KO samples, transcription factor footprints significantly increase around Adnp motifs, while nucleosome footprints decrease. 

#### Session information

```{r}
sessionInfo()
```

